scripts/japicmp.sh

```diff
@@ -15,43 +15,54 @@
+SCRIPT=$(basename "${BASH_SOURCE:-stidn}")
+
+if [ $# -lt 1 ] || [ $# -gt 3 ]; then
+  echo "# This script compares versions for binary backward compatibility."
+  echo "# Usage: ${SCRIPT} <old_version> (<new_version> (<group_id>))"
+  echo "# if <new_version> unspecified then comparison will be made againt local build"
+  echo "# if <group_id> unspecified then local dir gradle 'group' property will be used"
   exit 1
 fi
 
-MVN_REPO=$(mvn help:evaluate -Dexpression=settings.localRepository -q -DforceStdout)
+MVN_REPO="$(mvn help:evaluate -Dexpression=settings.localRepository -q -DforceStdout)"
 
-JAPICMP_VERSION=0.15.3
-JAR_FILE=$MVN_REPO/com/github/siom79/japicmp/japicmp/$JAPICMP_VERSION/japicmp-$JAPICMP_VERSION-jar-with-dependencies.jar
-if ! test -e "$JAR_FILE"; then
-    mvn -N dependency:get -DgroupId=com.github.siom79.japicmp -DartifactId=japicmp -Dversion=$JAPICMP_VERSION \
-        -Dtransitive=false -Dclassifier=jar-with-dependencies 1>&2
+JAPICMP_VERSION="0.15.3"
+JAR_DIR="${MVN_REPO}/com/github/siom79/japicmp/japicmp/${JAPICMP_VERSION}"
+JAR_FILE="${JAR_DIR}/japicmp-${JAPICMP_VERSION}-jar-with-dependencies.jar"
+if ! test -e "${JAR_FILE}"; then
+  mvn -N dependency:get \
+    -DgroupId=com.github.siom79.japicmp -DartifactId=japicmp -Dversion="${JAPICMP_VERSION}" \
+    -Dtransitive=false -Dclassifier=jar-with-dependencies 1>&2 || exit 1
 fi
 
-OLD_ST_VERSION=$1
-NEW_ST_VERSION=$2
-GROUP_ID=$3
-if [ -z "$GROUP_ID" ]; then
-  GROUP_ID=io.servicetalk
-fi
+OLD_ST_VERSION="${1}"
+NEW_ST_VERSION="${2:-$(./gradlew properties | grep '^version: ' | cut -f 2 -d ' ')}"
+GROUP_ID="${3:-$(./gradlew properties | grep '^group: ' | cut -f 2 -d ' ')}"
 GROUP_PATH=$(echo "$GROUP_ID" | tr '.' '/')
-BASEPATH=$MVN_REPO/$GROUP_PATH/
+BASEPATH="$MVN_REPO/$GROUP_PATH/"
 
 # All servicetalk modules except:
 # servicetalk-benchmarks, servicetalk-bom, servicetalk-examples, servicetalk-gradle-plugin-internal
-ARTIFACTS="$(ls -d -- */ | grep '^servicetalk-' | sed 's/.$//' | \
-  grep -v 'benchmark' | grep -v 'bom' | grep -v 'examples' | grep -v 'gradle-plugin-internal')"
-
-for ARTIFACT_ID in $ARTIFACTS
-do
-  mvn -N -U dependency:get -DgroupId=$GROUP_ID -DartifactId=$ARTIFACT_ID -Dversion=$OLD_ST_VERSION \
-      -Dtransitive=false >/dev/null
-  mvn -N -U dependency:get -DgroupId=$GROUP_ID -DartifactId=$ARTIFACT_ID -Dversion=$NEW_ST_VERSION \
-      -Dtransitive=false >/dev/null
-  java -jar "$JAR_FILE" -b --ignore-missing-classes \
-      --old $BASEPATH/$ARTIFACT_ID/$OLD_ST_VERSION/$ARTIFACT_ID-$OLD_ST_VERSION.jar \
-      --new $BASEPATH/$ARTIFACT_ID/$NEW_ST_VERSION/$ARTIFACT_ID-$NEW_ST_VERSION.jar
+ARTIFACTS="$(find servicetalk-* -type d -maxdepth 0 |
+  grep -v -- '-\(benchmarks\|bom\|examples\|gradle-plugin-internal\)$')"
+
+for ARTIFACT_ID in ${ARTIFACTS}; do
+  mvn -N -U dependency:get -DgroupId="${GROUP_ID}" -DartifactId="${ARTIFACT_ID}" \
+    -Dversion="${OLD_ST_VERSION}" -Dtransitive=false >/dev/null || continue
```

Comment:
in case of error, would we want anything reported at the end instead of skipping the error and going forward?

Reply:
The most likely cause of the error is a module that exists in the new version but not in the old version. The only way to do this "right" would be to process the bom lists from both the old and the new and the only diff the intersection but that seemed to be going too far. (We should look at the japicmp plugin instead for fancier use).

Reply:
I enhanced the script to now output a warning that it is skipping.